back_end/index.js

```diff
@@ -419,6 +419,163 @@ app.get('/fornecedor/:id', (req, res) => {
   console.log('get /fornecedor/' + fornecedorId + ' executado. Fornecedor retornado com sucesso!');
 });
   
+
+//alteração Erik e Marcos//
+
+app.get('/usuario', (req, res) => {
+  db.query('SELECT * FROM usuario', (err, results) => {
+    if (err) {
+      res.status(500).send('Erro retornando usuario: ' + err);
+      return;
+    }
+    res.json(results);
+  });
+  console.log('get /usuario executado. Usuarios retornados com sucesso!');
+});
+   
+/* POST */
+app.post('/usuario/create', (req, res) => {
+    const { login, nome, privilegios, senha } = req.body;
+    const query = `INSERT INTO usuario (login, nome, privilegios, senha) VALUES (?, ?, ?, ?)`;
+    const values = [login, nome, privilegios, senha];
+    db.query(query, values, (err, result) => {
```

Comment:
salvando a senha crua no banco direto do request body, sem criptografia

Reply:
Aqui temos uma questão que precisamos discutir e definir:
quando um usuário novo for incluído, a senha vai ficar vazia e depois o usuário novo grava a senha no primeiro logon? (isso é mais chato de fazer)
Ou quem for incluir o novo usuário já informa a senha do usuário novo no cadastro mesmo e depois o novo usuário altera a senha depois que ele fizer logon? (mais fácil)

Reply:
Esse tipo de implementação seria a adequada para o nosso caso?


```
app.post('/usuario/create', async (req, res) => {
    const { login, nome, privilegios, senha } = req.body;
    
    try {
        // Gerar um salt e hash a senha
        const saltRounds = 10; // Número de salt rounds
        const hashedPassword = await bcrypt.hash(senha, saltRounds); // Criptografa a senha

        const query = `INSERT INTO usuario (login, nome, privilegios, senha) VALUES (?, ?, ?, ?)`;
        const values = [login, nome, privilegios, hashedPassword]; // Usa a senha criptografada

        db.query(query, values, (err, result) => {
            if (err) {
                res.status(500).send('Erro criando usuario: ' + err);
                return;
            }
            const postId = result.insertId;
            db.query('SELECT * FROM usuario WHERE id_usuario = ?', postId, (err, result) => {
                if (err) {
                    res.status(500).send('Erro recuperando usuario criado: ' + err);
                    return;
                }
                res.status(201).json(result[0]);
            });
        });
        console.log('post executado. Usuario criado com sucesso!');
    } catch (error) {
        res.status(500).send('Erro ao hash da senha: ' + error);
    }
});
```

Reply:
Nao sei se no result da criação vai vir mesmo o id, eu conferiria isso num teste.
Tambem nao sei se o certo é passar um objeto ou um array como segundo argumento da função query, vc fez das duas formas isso me levanta suspeita, mas pode ser q de certo.

Ainda tem como melhorar:
```js
const INSERT_USER_QUERY = 'INSERT INTO usuario (login, nome, privilegios, senha) VALUES (?, ?, ?, ?)';
const SELECT_USER_QUERY = 'SELECT * FROM usuario WHERE id_usuario = ?';
const SALT_ROUNDS = 10;

app.post('/usuario', async (req, res) => {
    const { login, nome, privilegios, senha } = req.body;
    
    try {
        const hashedPassword = await bcrypt.hash(senha, SALT_ROUNDS);
        const values = [login, nome, privilegios, hashedPassword];

        db.query(INSERT_USER_QUERY, values, (err, result) => {
            if (err) {
                return res.status(500).send('Erro criando usuario: ' + err);
            }
            db.query(SELECT_USER_QUERY, result.insertId, (err, result) => {
                if (err) {
                    return  res.status(500).send('Erro recuperando usuario criado: ' + err);
                }
                res.status(201).json(result[0]);
            });
        });
        console.log('post executado. Usuario criado com sucesso!');
    } catch (error) {
        res.status(500).send('Erro ao fazer hash da senha: ' + error);
    }
});

Reply:
> Aqui temos uma questão que precisamos discutir e definir: quando um usuário novo for incluído, a senha vai ficar vazia e depois o usuário novo grava a senha no primeiro logon? (isso é mais chato de fazer) Ou quem for incluir o novo usuário já informa a senha do usuário novo no cadastro mesmo e depois o novo usuário altera a senha depois que ele fizer logon? (mais fácil)

Acredito que para fins práticos como vemos em diversos sistemas escolares, cursos e etc... seria interessante ter a famosa senha provisória do primeiro acesso como descreveu por último, o primeiro caso realmente é trabalhoso e não vejo necessidade disso e nem acho seguro, pois qualquer um que obtiver esse acesso poderá o pegar pra si, o único jeito de não ser inseguro seria se tivesse um dois fatores e acredito que não valha a pena a gente focar nisso no momento.